samples/RoutingSample.Web/Program.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text.RegularExpressions;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Constraints;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace RoutingSample.Web
+{
+    public class Program
+    {
+        private static readonly TimeSpan RegexMatchTimeout = TimeSpan.FromSeconds(10);
+
+        public static void Main(string[] args)
+        {
+            var host = new WebHostBuilder()
+                .UseKestrel()
+                .UseIISIntegration()
+                .ConfigureServices(services => services.AddRouting())
+                .Configure(app => app.UseRouter(routes =>
+                        {
+                            routes.DefaultHandler = new RouteHandler((httpContext) =>
+                            {
+                                var request = httpContext.Request;
+                                return httpContext.Response.WriteAsync($"Verb =  {request.Method.ToUpperInvariant()} - Path = {request.Path} - Route values - {string.Join(", ", httpContext.GetRouteData().Values)}");
+                            });
+
+                            routes.MapGet("api/get/{id}", (request, response, routeData) => response.WriteAsync($"API Get {routeData.Values["id"]}"))
+                                  .MapRoute("api/middleware", (IApplicationBuilder fork) => fork.Use((httpContext, next) => httpContext.Response.WriteAsync("Middleware!")))
```

Comment:
Had some discussion with @DamianEdwards and @Eilon lipton about these overloads (the ones that take `IApplicationBuilder`). We should **rename** these for 2.0.0 which would be a breaking change.

The problem is that lambda type inference can't work well because there are two overloads that take a single-parameter lambda.

Might want to double check with @Eilon - that could be either part of this PR or not (your choice).